server/src/main/java/org/elasticsearch/reservedstate/service/ReservedClusterStateService.java

```diff
@@ -101,17 +113,39 @@ public void process(String namespace, XContentParser parser, Consumer<Exception>
         try {
             stateChunk = stateChunkParser.apply(parser, null);
         } catch (Exception e) {
-            ErrorState errorState = new ErrorState(namespace, -1L, e, ReservedStateErrorMetadata.ErrorKind.PARSING);
-            saveErrorState(clusterService.state(), errorState);
-            logger.debug("error processing state change request for [{}] with the following errors [{}]", namespace, errorState);
+            errorStateForException(e, namespace, errorListener);
+            return;
+        }
 
-            errorListener.accept(
-                new IllegalStateException("Error processing state change request for " + namespace + ", errors: " + errorState, e)
-            );
+        if (stateChunk.state().isEmpty()) {
+            errorListener.accept(null);
             return;
         }
 
-        process(namespace, stateChunk, errorListener);
+        // After we have parsed the json content, we give each handler an opportunity to augment the data
```

Comment:
This refactoring was required in ReservedClusterStateService because the Ingest transform operations require NodeInfos, which is an async action. For this purpose, I introduced a preTransform step that runs with an async listener, allowing any handler to do additional fetching of data after we've parsed the JSON.